libs/checkpoint-redis/src/index.ts

```diff
@@ -0,0 +1,875 @@
+    // Apply TTL if configured
+    if (this.ttlConfig?.defaultTTL) {
+      await this.applyTTL(key);
+    }
+
+    return {
+      configurable: {
+        thread_id: threadId,
+        checkpoint_ns: checkpointNs,
+        checkpoint_id: checkpointId,
+      },
+    };
+  }
+
+  async *list(
+    config: RunnableConfig | null,
+    options?: CheckpointListOptions & { filter?: CheckpointMetadata }
+  ): AsyncGenerator<CheckpointTuple> {
+    await this.ensureIndexes();
+
+    // If filter is provided (even if empty), use search functionality
+    if (options?.filter !== undefined) {
+      // Check if we have null values in the filter which RediSearch can't handle
+      const hasNullFilter = Object.values(options.filter).some(
+        (v) => v === null
+      );
+
+      // Build search query
+      const queryParts: string[] = [];
+
+      // Add thread_id constraint if provided
+      if (config?.configurable?.thread_id) {
+        const threadId = config.configurable.thread_id.replace(
+          /[-.@]/g,
+          "\\$&"
+        );
+        queryParts.push(`(@thread_id:{${threadId}})`);
+      }
+
+      // Add checkpoint_ns constraint if provided
+      if (config?.configurable?.checkpoint_ns !== undefined) {
+        const checkpointNs = config.configurable.checkpoint_ns;
+        if (checkpointNs === "") {
+          // Empty string needs special handling in RediSearch
+          // We'll store it as "__empty__" in the index
+          queryParts.push(`(@checkpoint_ns:{__empty__})`);
+        } else {
+          const escapedNs = checkpointNs.replace(/[-.@]/g, "\\$&");
+          queryParts.push(`(@checkpoint_ns:{${escapedNs}})`);
+        }
+      }
+
+      // Skip metadata filters in search query when 'before' parameter is used
+      // We'll apply them after the before filtering to get correct results
+      if (!options?.before) {
+        // Add metadata filters (but skip null values)
+        for (const [key, value] of Object.entries(options.filter)) {
+          if (value === undefined) {
+            // Skip undefined filters
+          } else if (value === null) {
+            // Skip null values for RediSearch query, will handle in post-processing
+          } else if (typeof value === "string") {
+            // Don't escape, just wrap in braces for exact match
+            queryParts.push(`(@${key}:{${value}})`);
+          } else if (typeof value === "number") {
+            queryParts.push(`(@${key}:[${value} ${value}])`);
+          } else if (
+            typeof value === "object" &&
+            Object.keys(value).length === 0
+          ) {
+            // Skip empty objects
+          }
+        }
+      }
+
+      if (queryParts.length === 0) {
+        queryParts.push("*");
+      }
+
+      const query = queryParts.join(" ");
+      const limit = options?.limit ?? 10;
+
+      try {
+        // First get a larger set if we need to filter by 'before'
+        let searchLimit = options?.before ? 100 : limit;
```

Comment:
_q_: What will happen if a checkpoint ID specified `options.before` is outside the initial 100 checkpoint limit?

Reply:
Addressed, good catch.